pkg/storage/ingest/writer_test.go

```diff
@@ -0,0 +1,390 @@
+		writer, _ := createTestWriter(t, clusterAddr, topicName, createTestWriterConfig())
+
+		// Write to a non-existing partition.
+		err := writer.WriteSync(ctx, 100, tenantID, multiSeries, nil, mimirpb.API)
+		require.Error(t, err)
+	})
+
+	t.Run("should return an error and stop retrying sending a record once the write timeout expires", func(t *testing.T) {
+		t.Parallel()
+
+		cluster, clusterAddr := createTestCluster(t, numPartitions, topicName)
+		writerCfg := createTestWriterConfig()
+		writer, _ := createTestWriter(t, clusterAddr, topicName, writerCfg)
+
+		cluster.ControlKey(int16(kmsg.Produce), func(request kmsg.Request) (kmsg.Response, error, bool) {
+			// Keep failing every request.
+			cluster.KeepControl()
+			return nil, errors.New("mock error"), true
+		})
+
+		startTime := time.Now()
+		require.Equal(t, kgo.ErrRecordTimeout, writer.WriteSync(ctx, partitionID, tenantID, series1, nil, mimirpb.API))
+		elapsedTime := time.Since(startTime)
+
+		assert.Greater(t, elapsedTime, writerCfg.KafkaWriteTimeout/2)
+		assert.Less(t, elapsedTime, writerCfg.KafkaWriteTimeout*3) // High tolerance because the client does a backoff and timeout is evaluated after the backoff.
+	})
+
+	// This test documents how the Kafka client works. It's not what we ideally want, but it's how it works.
+	t.Run("should fail all buffered records and close the connection on timeout while waiting for Produce response", func(t *testing.T) {
+		t.Parallel()
+
+		cluster, clusterAddr := createTestCluster(t, numPartitions, topicName)
+		writerCfg := createTestWriterConfig()
+		writer, _ := createTestWriter(t, clusterAddr, topicName, writerCfg)
+
+		var (
+			firstRequest         = atomic.NewBool(true)
+			firstRequestReceived = make(chan struct{})
+		)
+
+		cluster.ControlKey(int16(kmsg.Produce), func(request kmsg.Request) (kmsg.Response, error, bool) {
+			if firstRequest.CompareAndSwap(true, false) {
+				// The produce request has been received by Kafka, so we can fire the next request.
+				close(firstRequestReceived)
+
+				// Inject a slowdown on the 1st Produce request received by Kafka.
+				// NOTE: the slowdown is 1s longer than the client timeout.
+				time.Sleep(writerCfg.KafkaWriteTimeout + writerRequestTimeoutOverhead + time.Second)
+			}
+
+			return nil, nil, false
+		})
+
+		wg := sync.WaitGroup{}
+		wg.Add(2)
+
+		// The 1st request is expected to fail because Kafka will take longer than the configured timeout.
+		runAsync(&wg, func() {
+			startTime := time.Now()
+			require.Equal(t, kgo.ErrRecordTimeout, writer.WriteSync(ctx, partitionID, tenantID, series1, nil, mimirpb.API))
+			elapsedTime := time.Since(startTime)
+
+			// It should take nearly the client's write timeout.
+			expectedElapsedTime := writerCfg.KafkaWriteTimeout + writerRequestTimeoutOverhead
+			tolerance := time.Second
+			assert.Greater(t, elapsedTime, expectedElapsedTime-tolerance)
+			assert.Less(t, elapsedTime, expectedElapsedTime+tolerance)
+		})
+
```

Comment:
can use 
```go
			assert.InEpsilon(t, expectedElapsedTime, elapsedTime, float64(time.Second))
```

Reply:
But I think we should use `InDelta()` given we pass the absolute max diff value.